RocketChat/Rocket.Chat
# feat: Add LDAP group validation strategy setting to channels and roles sync

## Proposed changes (including videos or screenshots)

- Added "Group membership validation strategy" settings to "Sync Channels" and "Sync Roles" LDAP sections -- they may enable a **faster** sync process when combined with a good LDAP search filter;

**Available strategies:**
 - **Apply filter for each group**: apply the LDAP user group filter for each group (key) defined in the LDAP group channel map. This is slower, but can be useful in case you need to use the `#{groupName}` replacement tag to define membership (e.g. when filtering by the `memberOf` field in groups);
 - **Apply filter once to get all memberships**: apply the LDAP user group filter once for each user. A given user will be considered a member of all groups returned by the LDAP search. This is a **faster** option that can be applied in case the `#{groupName}` replacement tag is not used by the filter (e.g. when filtering by the `member` field in groups).

## Issue(s)


## Steps to test or reproduce

The new "Group membership validation strategy" setting is available under both "Sync Channels" and "Sync Roles" sections in LDAP Premium settings. Both features should work just the same as in previous version when using the default "Apply filter for each group" search strategy or the new and faster "Apply filter once to get all memberships" strategy -- the only difference here is the amount of LDAP search requests triggered by RC.

<details>
  <summary><strong>Sample configuration</strong></summary>
  
<strong>Sample configuration for using the "Apply filter once to get all memberships" search strategy:</strong>
![Captura de tela de 2024-05-14 17-09-13](https://github.com/RocketChat/Rocket.Chat/assets/36537004/e06825ab-86e3-4866-bde6-ce32a97636e0)

<strong>Sample configuration for using the "Apply filter for each group" search strategy:</strong>
![Captura de tela de 2024-05-14 17-09-51](https://github.com/RocketChat/Rocket.Chat/assets/36537004/f20377e5-40ce-4b3e-8471-c7accab8fbfa)
  
</details>

> [!CAUTION]
> Switching to the new and faster "Apply filter once to get all memberships" search strategy may not work with the currently configured LDAP search filter!
> When switching to the new strategy, make sure to update the user group filter so as to get **all groups at once** in a single query with it (be sure not to use the `#{groupName}` replacement tag since it is not supported by this new strategy)

## Further comments

[CORE-402](https://rocketchat.atlassian.net/browse/CORE-402)

[CORE-402]: https://rocketchat.atlassian.net/browse/CORE-402?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
